File: src/cli/parse.ts
/**
 * Hand-rolled argv parser. Avoids external deps so the CLI is dependency-free
 * (matters for `npx claude-profiles`-style cold starts) and so the parsing
 * surface is fully reviewable in one file.
 *
 * Conventions:
 *  - First non-flag token is the verb (R29). `--help`/`--version` short-
 *    circuit verb dispatch.
 *  - Global flags (`--json`, `--cwd=<path>`, `--on-drift=<choice>`,
 *    `--no-color`, `--help`, `--version`) may appear anywhere in argv —
 *    before or after the verb.
 *  - Verb-specific flags (`--description=<txt>` for `new`, `--pre-commit-warn`
 *    for `drift`) are scoped per verb.
 *  - Positional args after the verb are required (e.g. `use <name>`); we
 *    return ParseError naming the missing arg, never throw.
 *  - Unknown verbs / flags / values are ParseErrors with exit-code-1 semantics.
 */

import type { GlobalOptions, OnDriftFlag, ParsedInvocation } from "./types.js";

export interface ParseError {
  ok: false;
  /** Human-readable error suitable for stderr. */
  message: string;
  /** True iff the user typed `--help` or `--version` (no args required). */
  helpRequested: boolean;
}

export type ParseResult =
  | { ok: true; invocation: ParsedInvocation }
  | ParseError;

const VERBS = new Set([
  "init",
  "list",
  "use",
  "status",
  "drift",
  "diff",
  "new",
  "validate",
  "sync",
  "hook",
  "help",
]);

const ON_DRIFT_VALUES: ReadonlySet<OnDriftFlag> = new Set<OnDriftFlag>([
  "discard",
  "persist",
  "abort",
]);

/**
 * Parse argv (without the leading `node` + script path — caller slices off
 * `process.argv.slice(2)`). Returns either a typed ParsedInvocation or a
 * structured error.
 *
 * `defaultCwd` is the cwd to use when `--cwd=` is not passed; the bin entry
 * supplies `process.cwd()`.
 */
export function parseArgs(argv: ReadonlyArray<string>, defaultCwd: string): ParseResult {
  // Collect global flags + extract verb + remaining tokens in a single pass.
  const tokens = [...argv];
  const global: GlobalOptions = {
    json: false,
    cwd: defaultCwd,
    onDrift: null,
    noColor: false,
    quiet: false,
    waitMs: null,
  };

  // Side-channel for help/version short-circuit. We still want to parse the
  // verb (so `claude-profiles use --help` shows use-specific help), but we
  // surface the request via the Command itself.
  let helpFlagSeen = false;
  let versionFlagSeen = false;

  // First pass: pull global flags out of the token stream. We do this BEFORE
  // verb dispatch so `--cwd=/tmp use foo` and `use foo --cwd=/tmp` both work.
  const verbAndArgs: string[] = [];
  for (let i = 0; i < tokens.length; i++) {
    const t = tokens[i]!;
    if (t === "--json") {
      global.json = true;
    } else if (t === "--help" || t === "-h") {
      helpFlagSeen = true;
    } else if (t === "--version" || t === "-V") {
      versionFlagSeen = true;
    } else if (t === "--cwd") {
      const next = tokens[i + 1];
      if (next === undefined || next.startsWith("-")) {
        return parseError(`--cwd requires a path argument`);
      }
      // Empty-string is rejected here as well as in the `--cwd=` branch — a
      // bare `""` arg silently running against cwd="" would later look up
      // `.claude-profiles/` in whatever directory the OS normalises "" to.
      if (next === "") return parseError(`--cwd requires a non-empty path`);
      global.cwd = next;
      i++;
    } else if (t.startsWith("--cwd=")) {
      global.cwd = t.slice("--cwd=".length);
      if (global.cwd === "") return parseError(`--cwd requires a non-empty path`);
    } else if (t === "--on-drift") {
      const next = tokens[i + 1];
      if (next === undefined) return parseError(`--on-drift requires a value (discard|persist|abort)`);
      const flag = parseOnDrift(next);
      if (!flag) return parseError(`--on-drift must be discard|persist|abort, got "${next}"`);
      global.onDrift = flag;
      i++;
    } else if (t.startsWith("--on-drift=")) {
      const v = t.slice("--on-drift=".length);
      const flag = parseOnDrift(v);
      if (!flag) return parseError(`--on-drift must be discard|persist|abort, got "${v}"`);
      global.onDrift = flag;
    } else if (t === "--no-color") {
      // Additive with NO_COLOR env: the flag turns colour off even when env
      // is unset. Threaded through dispatch into createStyle.
      global.noColor = true;
    } else if (t === "--quiet" || t === "-q") {
      // Silences human print()/warn() while preserving error() and exit code.
      // Mutually-exclusive enforcement happens after the full pass so users
      // see one consolidated error regardless of arg order.
      global.quiet = true;
    } else if (t === "--wait") {
      // yd8 / AC-4: bare --wait → 30 second default budget. Picked to be
      // long enough that a typical short swap (sub-second) clears, while
      // capping accidental hangs at well under the impatient-user threshold.
      global.waitMs = 30_000;
    } else if (t.startsWith("--wait=")) {
      const v = t.slice("--wait=".length);
      const seconds = Number.parseFloat(v);
      if (!Number.isFinite(seconds) || seconds < 0) {
        return parseError(`--wait must be a non-negative number of seconds; got "${v}"`);
      }
      global.waitMs = Math.round(seconds * 1000);
    } else {
      verbAndArgs.push(t);
    }
  }

  // --quiet and --json are mutually exclusive: --json already silences human
  // output, and a script that asks for both is signalling unclear intent —
  // surface the conflict at parse time rather than silently picking one.
  if (global.quiet && global.json) {
    return parseError(`--quiet and --json are mutually exclusive`);
  }

  // Version short-circuit beats verb dispatch (R29 doesn't list `--version`
  // as a verb but it's a universal CLI affordance). Fires regardless of
  // whether a verb is also present — silently consuming `--version` when a
  // verb is supplied (`claude-profiles list --version`) hides the user's
  // intent. The verb form `--help` keeps the verb (renders verb-specific
  // help), but `--version` always means "print version and exit".
  if (versionFlagSeen) {
    return ok({ command: { kind: "version" }, global });
  }

  // Help with no verb -> top-level help.
  if (verbAndArgs.length === 0) {
    if (helpFlagSeen) return ok({ command: { kind: "help", verb: null }, global });
    return parseError(
      `missing command; run "claude-profiles --help" for usage`,
      true,
    );
  }

  const verb = verbAndArgs[0]!;
  const rest = verbAndArgs.slice(1);

  if (!VERBS.has(verb)) {
    return parseError(`unknown command "${verb}"; run "claude-profiles --help" for usage`);
  }

  // `help <verb>` is equivalent to `<verb> --help`. Both produce a help
  // command for the given verb.
  if (verb === "help") {
    if (rest.length === 0) return ok({ command: { kind: "help", verb: null }, global });
    if (rest.length > 1) return parseError(`help takes at most one argument; got "${rest.join(" ")}"`);
    return ok({ command: { kind: "help", verb: rest[0]! }, global });
  }

  if (helpFlagSeen) {
    return ok({ command: { kind: "help", verb }, global });
  }

  // Per-verb dispatch.
  switch (verb) {
    case "init": {
      let starter = "default";
      let seed = true;
      let hook = true;
      const positional: string[] = [];
      for (let i = 0; i < rest.length; i++) {
        const t = rest[i]!;
        if (t === "--no-seed") seed = false;
        else if (t === "--no-hook") hook = false;
        else if (t === "--starter") {
          const next = rest[i + 1];
          // Treat any leading `-` as a missing value: profile names never
          // start with `-`, so `init --starter -force` (a typo) would
          // otherwise silently bind `starter = "-force"`.
          if (next === undefined || next.startsWith("-")) {
            return parseError(`--starter requires a value`);
          }
          starter = next;
          i++;
        } else if (t.startsWith("--starter=")) {
          starter = t.slice("--starter=".length);
          if (starter === "") return parseError(`--starter requires a non-empty value`);
        } else if (t.startsWith("--")) {
          return parseError(`init: unknown flag "${t}"`);
        } else {
          positional.push(t);
        }
      }
      if (positional.length > 0) {
        return parseError(`init takes no positional arguments; got "${positional.join(" ")}"`);
      }
      return ok({ command: { kind: "init", starter, seed, hook }, global });
    }

    case "list":
      if (rest.length > 0) return parseError(`list takes no arguments; got "${rest.join(" ")}"`);
      return ok({ command: { kind: "list" }, global });

    case "status":
      if (rest.length > 0) return parseError(`status takes no arguments; got "${rest.join(" ")}"`);
      return ok({ command: { kind: "status" }, global });

    case "sync":
      if (rest.length > 0) return parseError(`sync takes no arguments; got "${rest.join(" ")}"`);
      return ok({ command: { kind: "sync" }, global });

    case "use": {
      if (rest.length === 0) return parseError(`use requires a profile name`);
      if (rest.length > 1) return parseError(`use takes one argument; got "${rest.join(" ")}"`);
      return ok({ command: { kind: "use", profile: rest[0]! }, global });
    }

    case "drift": {
      let preCommitWarn = false;
      let verbose = false;
      let preview = false;
      const positional: string[] = [];
      for (const t of rest) {
        if (t === "--pre-commit-warn") preCommitWarn = true;
        else if (t === "--verbose") verbose = true;
        else if (t === "--preview") preview = true;
        else if (t.startsWith("--")) return parseError(`drift: unknown flag "${t}"`);
        else positional.push(t);
      }
      if (positional.length > 0) return parseError(`drift takes no positional arguments; got "${positional.join(" ")}"`);
      return ok({ command: { kind: "drift", preCommitWarn, verbose, preview }, global });
    }

    case "diff": {
      let preview = false;
      const positional: string[] = [];
      for (const t of rest) {
        if (t === "--preview") preview = true;
        else if (t.startsWith("--")) return parseError(`diff: unknown flag "${t}"`);
        else positional.push(t);
      }
      if (positional.length === 0) return parseError(`diff requires at least one profile name`);
      if (positional.length > 2) return parseError(`diff takes one or two profile names; got "${positional.join(" ")}"`);
      return ok({
        command: { kind: "diff", a: positional[0]!, b: positional[1] ?? null, preview },
        global,
      });
    }

    case "new": {
      let description: string | null = null;
      const positional: string[] = [];
      for (let i = 0; i < rest.length; i++) {
        const t = rest[i]!;
        if (t === "--description") {
          const next = rest[i + 1];
          if (next === undefined || next.startsWith("--")) {
            return parseError(`--description requires a value`);
          }
          description = next;
          i++;
        } else if (t.startsWith("--description=")) {
          description = t.slice("--description=".length);
        } else if (t.startsWith("--")) {
          return parseError(`new: unknown flag "${t}"`);
        } else {
          positional.push(t);
        }
      }
      if (positional.length === 0) return parseError(`new requires a profile name`);
      if (positional.length > 1) return parseError(`new takes one positional argument; got "${positional.join(" ")}"`);
      return ok({
        command: { kind: "new", profile: positional[0]!, description },
        global,
      });
    }

    case "validate": {
      let brief = false;
      const positional: string[] = [];
      for (const t of rest) {
        if (t === "--brief") brief = true;
        else if (t.startsWith("--")) return parseError(`validate: unknown flag "${t}"`);
        else positional.push(t);
      }
      if (positional.length > 1) return parseError(`validate takes at most one profile name; got "${positional.join(" ")}"`);
      return ok({
        command: { kind: "validate", profile: positional[0] ?? null, brief },
        global,
      });
    }

    case "hook": {
      let force = false;
      const positional: string[] = [];
      for (const t of rest) {
        if (t === "--force") force = true;
        else if (t.startsWith("--")) return parseError(`hook: unknown flag "${t}"`);
        else positional.push(t);
      }
      if (positional.length === 0) return parseError(`hook requires an action (install|uninstall)`);
      if (positional.length > 1) return parseError(`hook takes one positional argument; got "${positional.join(" ")}"`);
      const action = positional[0]!;
      if (action !== "install" && action !== "uninstall") {
        return parseError(`hook action must be install|uninstall, got "${action}"`);
      }
      // `--force` is only meaningful for `install` (overwrite a foreign
      // hook). Silently accepting it on `uninstall` invites scripts that
      // think they're forcing removal — but our policy is to NEVER remove
      // a non-matching hook. Reject explicitly so the contract is clear.
      if (action === "uninstall" && force) {
        return parseError(`hook uninstall does not accept --force (a non-matching hook is never removed)`);
      }
      return ok({ command: { kind: "hook", action, force }, global });
    }

    default: {
      // Should be unreachable thanks to VERBS.has() guard above.
      const _exhaustive: string = verb;
      return parseError(`unknown command "${_exhaustive}"`);
    }
  }
}

function parseOnDrift(value: string): OnDriftFlag | null {
  if (ON_DRIFT_VALUES.has(value as OnDriftFlag)) return value as OnDriftFlag;
  return null;
}

function ok(invocation: ParsedInvocation): ParseResult {
  return { ok: true, invocation };
}

function parseError(message: string, helpRequested = false): ParseError {
  return { ok: false, message, helpRequested };
}

File: internal/cli/parse.go
// Hand-rolled argv parser. Mirrors src/cli/parse.ts conventions:
//   - first non-flag token is the verb
//   - global flags (--json, --cwd, --on-drift, --no-color, --quiet, --wait)
//     may appear anywhere in argv, before or after the verb
//   - verb-specific flags scoped per verb
//   - positional args after the verb are required where named (we return
//     ParseError naming the missing arg, never panic)
//
// Cobra is intentionally NOT used: cobra's parsing rules differ from the TS
// hand-rolled parser (flag positioning, error messages, --version
// short-circuit semantics) and the cross-language IV harness compares both
// surfaces. Keeping the parser surface tight to TS makes parity test work
// once instead of twice.
package cli

import (
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"

	"github.com/htxryan/claude-code-config-profiles/internal/drift"
)

// ParseError carries the human-readable message and a help-requested flag.
type ParseError struct {
	Message       string
	HelpRequested bool
}

func (e *ParseError) Error() string { return e.Message }

// ParseResult is either ok (with a typed invocation) or an error.
type ParseResult struct {
	Ok         bool
	Invocation ParsedInvocation
	Err        *ParseError
}

var verbs = map[string]bool{
	"init":        true,
	"list":        true,
	"use":         true,
	"status":      true,
	"drift":       true,
	"diff":        true,
	"new":         true,
	"validate":    true,
	"sync":        true,
	"hook":        true,
	"doctor":      true,
	"completions": true,
	"help":        true,
	// `hello` is intentionally hidden from help but parses + dispatches.
	"hello": true,
}

var completionShells = map[string]CompletionShell{
	"bash": ShellBash,
	"zsh":  ShellZsh,
	"fish": ShellFish,
}

// ParseArgs parses argv (without the leading binary name) plus an optional
// default cwd (production callers pass os.Getwd()).
func ParseArgs(argv []string, defaultCwd string) ParseResult {
	g := GlobalOptions{Cwd: defaultCwd}
	helpFlag := false
	versionFlag := false

	var verbAndArgs []string
	for i := 0; i < len(argv); i++ {
		t := argv[i]
		switch {
		case t == "--json":
			g.JSON = true
		case t == "--help" || t == "-h":
			helpFlag = true
		case t == "--version" || t == "-V":
			versionFlag = true
		case t == "--cwd":
			next, ok := nextArg(argv, i)
			if !ok || strings.HasPrefix(next, "-") {
				return errResult("--cwd requires a path argument", false)
			}
			if next == "" {
				return errResult("--cwd requires a non-empty path", false)
			}
			g.Cwd = next
			i++
		case strings.HasPrefix(t, "--cwd="):
			v := strings.TrimPrefix(t, "--cwd=")
			if v == "" {
				return errResult("--cwd requires a non-empty path", false)
			}
			g.Cwd = v
		case t == "--on-drift":
			next, ok := nextArg(argv, i)
			if !ok {
				return errResult("--on-drift requires a value (discard|persist|abort)", false)
			}
			c, ok := parseOnDrift(next)
			if !ok {
				return errResult(fmt.Sprintf(`--on-drift must be discard|persist|abort, got %q`, next), false)
			}
			g.OnDrift = c
			i++
		case strings.HasPrefix(t, "--on-drift="):
			v := strings.TrimPrefix(t, "--on-drift=")
			c, ok := parseOnDrift(v)
			if !ok {
				return errResult(fmt.Sprintf(`--on-drift must be discard|persist|abort, got %q`, v), false)
			}
			g.OnDrift = c
		case t == "--no-color":
			g.NoColor = true
		case t == "--quiet" || t == "-q":
			g.Quiet = true
		case t == "--wait":
			g.WaitMs = 30_000
			g.WaitSet = true
		case strings.HasPrefix(t, "--wait="):
			v := strings.TrimPrefix(t, "--wait=")
			seconds, err := strconv.ParseFloat(v, 64)
			if err != nil || seconds < 0 {
				return errResult(fmt.Sprintf(`--wait must be a non-negative number of seconds; got %q`, v), false)
			}
			g.WaitMs = int64(math.Round(seconds * 1000))
			g.WaitSet = true
		case t == "--non-interactive":
			g.NonInteractive = true
		default:
			verbAndArgs = append(verbAndArgs, t)
		}
	}

	// CI=true env auto-detection (only when the flag isn't already set).
	if !g.NonInteractive && os.Getenv("CI") == "true" {
		g.NonInteractive = true
	}

	if g.Quiet && g.JSON {
		return errResult("--quiet and --json are mutually exclusive", false)
	}

	if versionFlag {
		return ok(ParsedInvocation{Command: Command{Kind: KindVersion}, Global: g})
	}

	if len(verbAndArgs) == 0 {
		if helpFlag {
			return ok(ParsedInvocation{Command: Command{Kind: KindHelp}, Global: g})
		}
		return errResult(`missing command; run "c3p --help" for usage`, true)
	}

	verb := verbAndArgs[0]
	rest := verbAndArgs[1:]
	if !verbs[verb] {
		return errResult(fmt.Sprintf(`unknown command %q; run "c3p --help" for usage`, verb), false)
	}

	if verb == "help" {
		if len(rest) == 0 {
			return ok(ParsedInvocation{Command: Command{Kind: KindHelp}, Global: g})
		}
		if len(rest) > 1 {
			return errResult(fmt.Sprintf(`help takes at most one argument; got %q`, strings.Join(rest, " ")), false)
		}
		return ok(ParsedInvocation{Command: Command{Kind: KindHelp, HelpVerb: rest[0]}, Global: g})
	}

	if helpFlag {
		return ok(ParsedInvocation{Command: Command{Kind: KindHelp, HelpVerb: verb}, Global: g})
	}

	switch verb {
	case "init":
		return parseInit(rest, g)
	case "list":
		if len(rest) > 0 {
			return errResult(fmt.Sprintf(`list takes no arguments; got %q`, strings.Join(rest, " ")), false)
		}
		return ok(ParsedInvocation{Command: Command{Kind: KindList}, Global: g})
	case "status":
		if len(rest) > 0 {
			return errResult(fmt.Sprintf(`status takes no arguments; got %q`, strings.Join(rest, " ")), false)
		}
		return ok(ParsedInvocation{Command: Command{Kind: KindStatus}, Global: g})
	case "sync":
		if len(rest) > 0 {
			return errResult(fmt.Sprintf(`sync takes no arguments; got %q`, strings.Join(rest, " ")), false)
		}
		return ok(ParsedInvocation{Command: Command{Kind: KindSync}, Global: g})
	case "hello":
		if len(rest) > 0 {
			return errResult(fmt.Sprintf(`hello takes no arguments; got %q`, strings.Join(rest, " ")), false)
		}
		return ok(ParsedInvocation{Command: Command{Kind: KindHello}, Global: g})
	case "use":
		if len(rest) == 0 {
			return errResult("use requires a profile name", false)
		}
		if len(rest) > 1 {
			return errResult(fmt.Sprintf(`use takes one argument; got %q`, strings.Join(rest, " ")), false)
		}
		return ok(ParsedInvocation{Command: Command{Kind: KindUse, Profile: rest[0]}, Global: g})
	case "drift":
		return parseDrift(rest, g)
	case "diff":
		return parseDiff(rest, g)
	case "new":
		return parseNew(rest, g)
	case "validate":
		return parseValidate(rest, g)
	case "hook":
		return parseHook(rest, g)
	case "doctor":
		return parseDoctor(rest, g)
	case "completions":
		return parseCompletions(rest, g)
	}

	return errResult(fmt.Sprintf("unreachable: unknown verb %q", verb), false)
}

func parseInit(rest []string, g GlobalOptions) ParseResult {
	starter := "default"
	seed := true
	hook := true
	var positional []string
	for i := 0; i < len(rest); i++ {
		t := rest[i]
		switch {
		case t == "--no-seed":
			seed = false
		case t == "--no-hook":
			hook = false
		case t == "--starter":
			next, ok := nextArg(rest, i)
			if !ok || strings.HasPrefix(next, "-") {
				return errResult("--starter requires a value", false)
			}
			starter = next
			i++
		case strings.HasPrefix(t, "--starter="):
			v := strings.TrimPrefix(t, "--starter=")
			if v == "" {
				return errResult("--starter requires a non-empty value", false)
			}
			starter = v
		case strings.HasPrefix(t, "--"):
			return errResult(fmt.Sprintf(`init: unknown flag %q`, t), false)
		default:
			positional = append(positional, t)
		}
	}
	if len(positional) > 0 {
		return errResult(fmt.Sprintf(`init takes no positional arguments; got %q`, strings.Join(positional, " ")), false)
	}
	return ok(ParsedInvocation{
		Command: Command{Kind: KindInit, Starter: starter, Seed: seed, Hook: hook},
		Global:  g,
	})
}

func parseDrift(rest []string, g GlobalOptions) ParseResult {
	var preCommitWarn, verbose, preview bool
	var positional []string
	for _, t := range rest {
		switch {
		case t == "--pre-commit-warn":
			preCommitWarn = true
		case t == "--verbose":
			verbose = true
		case t == "--preview":
			preview = true
		case strings.HasPrefix(t, "--"):
			return errResult(fmt.Sprintf(`drift: unknown flag %q`, t), false)
		default:
			positional = append(positional, t)
		}
	}
	if len(positional) > 0 {
		return errResult(fmt.Sprintf(`drift takes no positional arguments; got %q`, strings.Join(positional, " ")), false)
	}
	return ok(ParsedInvocation{
		Command: Command{Kind: KindDrift, PreCommitWarn: preCommitWarn, Verbose: verbose, Preview: preview},
		Global:  g,
	})
}

func parseDiff(rest []string, g GlobalOptions) ParseResult {
	var preview bool
	var positional []string
	for _, t := range rest {
		switch {
		case t == "--preview":
			preview = true
		case strings.HasPrefix(t, "--"):
			return errResult(fmt.Sprintf(`diff: unknown flag %q`, t), false)
		default:
			positional = append(positional, t)
		}
	}
	if len(positional) == 0 {
		return errResult("diff requires at least one profile name", false)
	}
	if len(positional) > 2 {
		return errResult(fmt.Sprintf(`diff takes one or two profile names; got %q`, strings.Join(positional, " ")), false)
	}
	cmd := Command{Kind: KindDiff, A: positional[0], Preview: preview}
	if len(positional) == 2 {
		cmd.B = positional[1]
	}
	return ok(ParsedInvocation{Command: cmd, Global: g})
}

func parseNew(rest []string, g GlobalOptions) ParseResult {
	description := ""
	hasDescription := false
	var positional []string
	for i := 0; i < len(rest); i++ {
		t := rest[i]
		switch {
		case t == "--description":
			next, ok := nextArg(rest, i)
			if !ok || strings.HasPrefix(next, "--") {
				return errResult("--description requires a value", false)
			}
			description = next
			hasDescription = true
			i++
		case strings.HasPrefix(t, "--description="):
			description = strings.TrimPrefix(t, "--description=")
			hasDescription = true
		case strings.HasPrefix(t, "--"):
			return errResult(fmt.Sprintf(`new: unknown flag %q`, t), false)
		default:
			positional = append(positional, t)
		}
	}
	if len(positional) == 0 {
		return errResult("new requires a profile name", false)
	}
	if len(positional) > 1 {
		return errResult(fmt.Sprintf(`new takes one positional argument; got %q`, strings.Join(positional, " ")), false)
	}
	cmd := Command{Kind: KindNew, Profile: positional[0]}
	if hasDescription {
		cmd.Description = description
	}
	return ok(ParsedInvocation{Command: cmd, Global: g})
}

func parseValidate(rest []string, g GlobalOptions) ParseResult {
	var brief bool
	var positional []string
	for _, t := range rest {
		switch {
		case t == "--brief":
			brief = true
		case strings.HasPrefix(t, "--"):
			return errResult(fmt.Sprintf(`validate: unknown flag %q`, t), false)
		default:
			positional = append(positional, t)
		}
	}
	if len(positional) > 1 {
		return errResult(fmt.Sprintf(`validate takes at most one profile name; got %q`, strings.Join(positional, " ")), false)
	}
	cmd := Command{Kind: KindValidate, Brief: brief}
	if len(positional) == 1 {
		cmd.ValidateProfile = positional[0]
	}
	return ok(ParsedInvocation{Command: cmd, Global: g})
}

func parseHook(rest []string, g GlobalOptions) ParseResult {
	var force bool
	var positional []string
	for _, t := range rest {
		switch {
		case t == "--force":
			force = true
		case strings.HasPrefix(t, "--"):
			return errResult(fmt.Sprintf(`hook: unknown flag %q`, t), false)
		default:
			positional = append(positional, t)
		}
	}
	if len(positional) == 0 {
		return errResult("hook requires an action (install|uninstall)", false)
	}
	if len(positional) > 1 {
		return errResult(fmt.Sprintf(`hook takes one positional argument; got %q`, strings.Join(positional, " ")), false)
	}
	action := positional[0]
	if action != "install" && action != "uninstall" {
		return errResult(fmt.Sprintf(`hook action must be install|uninstall, got %q`, action), false)
	}
	if action == "uninstall" && force {
		return errResult("hook uninstall does not accept --force (a non-matching hook is never removed)", false)
	}
	return ok(ParsedInvocation{
		Command: Command{Kind: KindHook, HookAction: HookAction(action), Force: force},
		Global:  g,
	})
}

func parseDoctor(rest []string, g GlobalOptions) ParseResult {
	var positional []string
	for _, t := range rest {
		if strings.HasPrefix(t, "--") {
			return errResult(fmt.Sprintf(`doctor: unknown flag %q`, t), false)
		}
		positional = append(positional, t)
	}
	if len(positional) > 0 {
		return errResult(fmt.Sprintf(`doctor takes no positional arguments; got %q`, strings.Join(positional, " ")), false)
	}
	return ok(ParsedInvocation{Command: Command{Kind: KindDoctor}, Global: g})
}

func parseCompletions(rest []string, g GlobalOptions) ParseResult {
	var positional []string
	for _, t := range rest {
		if strings.HasPrefix(t, "--") {
			return errResult(fmt.Sprintf(`completions: unknown flag %q`, t), false)
		}
		positional = append(positional, t)
	}
	if len(positional) == 0 {
		return errResult("completions requires a shell (bash|zsh|fish)", false)
	}
	if len(positional) > 1 {
		return errResult(fmt.Sprintf(`completions takes one positional argument; got %q`, strings.Join(positional, " ")), false)
	}
	shell, ok := completionShells[positional[0]]
	if !ok {
		return errResult(fmt.Sprintf(`completions shell must be bash|zsh|fish, got %q`, positional[0]), false)
	}
	return okResult(ParsedInvocation{Command: Command{Kind: KindCompletions, Shell: shell}, Global: g})
}

func parseOnDrift(s string) (drift.GateChoice, bool) {
	switch s {
	case "discard":
		return drift.GateChoiceDiscard, true
	case "persist":
		return drift.GateChoicePersist, true
	case "abort":
		return drift.GateChoiceAbort, true
	}
	return "", false
}

func nextArg(argv []string, i int) (string, bool) {
	if i+1 >= len(argv) {
		return "", false
	}
	return argv[i+1], true
}

func ok(inv ParsedInvocation) ParseResult     { return ParseResult{Ok: true, Invocation: inv} }
func okResult(inv ParsedInvocation) ParseResult { return ok(inv) }

func errResult(msg string, helpRequested bool) ParseResult {
	return ParseResult{Ok: false, Err: &ParseError{Message: msg, HelpRequested: helpRequested}}
}
